data/import_data.py

```diff
@@ -0,0 +1,191 @@
+    return begin, end, what, who, room, building
+
+
+def cards_for_course_and_part(course, part_id):
+    lecture_ids = [int(lecture.attrib['id']) for lecture in course_lectures(course)]
+    lecture_predicate = " or ".join(["lecture/@ref={}".format(lecture_id) for lecture_id in lecture_ids])
+    return course.xpath("/definitions/cards/card[part/@ref={}][{}]".format(part_id, lecture_predicate))
+
+
+@functools.lru_cache(1000)
+def unique_course_time_slots(course):
+    return {frozenset({card_data(card) for card in cards_for_course_and_part(course, part_id)
+                       if is_card_valid(card)})
+            for part_id in part_ids_for_course(course)}
+
+
+@functools.lru_cache(1000)
+def jsonable_time_slots(course):
+    keys = ("od", "do", "typ", "vyucujici", "mistnost", "budova")
+    time_slots_with_possible_garbage = ([dict(zip(keys, values)) for values in variant if values[0] != 0]
+                                        for variant in unique_course_time_slots(course))
+    return [time_slot_list for time_slot_list in time_slots_with_possible_garbage if len(time_slot_list) > 0]
+
+
+@functools.lru_cache(1000)
+def translate_subject_code(timetable, subject_code):
+    exceptions = {"01VAM": "VAM/B", "01VAMB": "VAM/B",
+                  "01LIP": "LIP/B", "01LIPB": "LIP/B",
+                  "17EHJE": "ZEH",
+                  "17DEZ": "DEZ", "17PTA": "DEZ",
+                  "17JBEZ": "JBEZ/", "17JABE": "JBEZ/", "17ZBJE": "JBEZ/",
+                  "17ELZ": "TCJ1", "17TCJ1": "TCJ1",
+                  "17PRE": "NRE"
+                  }
+    if subject_code in exceptions:
+        return exceptions[subject_code]
+    elif subject_code.lstrip(string.digits) in subject_codes(timetable):
+        return subject_code.lstrip(string.digits)
+    else:
+        return None
+
+
+@functools.lru_cache(maxsize=1000)
+def get_timetable_data(timetable, subject_code):
+    possible_match = translate_subject_code(timetable, subject_code)
+    return [] if possible_match is None else jsonable_time_slots(get_course(timetable, possible_match))
+
+
+def guess_semester_code():
+    now = datetime.datetime.now()
+    return "B{}{}".format(str(now.year % 100), 2 if now.month < 8 else 1)
+
+
+def guess_timetable_uri():
+    now = datetime.datetime.now()
+    year = now.year % 100 - (1 if now.month < 8 else 0)
+    semester = "letni" if now.month < 8 else "zimni"
+    return "https://rozvrh.fjfi.cvut.cz/timetable/{}-{}-{}.xml".format(year, year + 1, semester)
+
+
+def extract_subject_data(subject, timetable):
+    subject_code = subject.xpath("string(kod)")
+    return {
+        "predmet_id": int(subject.xpath("string(predmet_id)")),
+        "kod": subject_code,
+        "nazev": subject.xpath("string(nazev)"),
+        "kredity": int(subject.xpath("string(kredity)")),
+        "rozsah": subject.xpath("string(rozsah)"),
+        "rozvrhy": get_timetable_data(timetable, subject_code),
```

Comment:
`get_timetable_data` je pro mě black-box. Snažil jsem se to přečíst, ale je to trochu moc komplikované (což není výtka, ten problém je komplikovaný). Jediné co by bylo fajn je, napsat krátkou dokumentaci k funkcím, které jsou hlavní building blocks (např. právě `get_timetable_data`).